libs/core/algorithms/include/hpx/parallel/run_on_all.hpp

```diff
@@ -19,70 +23,179 @@
 #include <hpx/executors/parallel_executor.hpp>
 #include <hpx/functional/experimental/scope_exit.hpp>
 #include <hpx/parallel/algorithms/for_loop_reduction.hpp>
+#include <hpx/parallel/execution.hpp>
 
 #include <cstddef>
 #include <type_traits>
 
 namespace hpx::experimental {
 
-    template <typename T, typename Op, typename F, typename... Ts>
-    void run_on_all(std::size_t num_tasks,
+    /// \brief Run a function on all available worker threads with reduction support
+    /// \tparam ExPolicy The execution policy type
+    /// \tparam T The reduction value type
+    /// \tparam Op The reduction operation type
+    /// \tparam F The function type to execute
+    /// \tparam Ts Additional argument types
+    /// \param policy The execution policy to use
+    /// \param num_tasks The number of tasks to create
+    /// \param r The reduction helper
+    /// \param f The function to execute
+    /// \param ts Additional arguments to pass to the function
+    template <typename ExPolicy, typename T, typename Op, typename F,
+        typename... Ts>
+    decltype(auto) run_on_all(ExPolicy&& policy, std::size_t num_tasks,
         hpx::parallel::detail::reduction_helper<T, Op>&& r, F&& f, Ts&&... ts)
     {
-        // force using index_queue scheduler with given amount of threads
+        static_assert(hpx::is_execution_policy_v<ExPolicy>,
+            "hpx::is_execution_policy_v<ExPolicy>");
+
+        // Configure executor with proper scheduling hints
         hpx::threads::thread_schedule_hint hint;
         hint.sharing_mode(
             hpx::threads::thread_sharing_hint::do_not_share_function);
+
         auto exec = hpx::execution::experimental::with_processing_units_count(
             hpx::execution::parallel_executor(
                 hpx::threads::thread_priority::bound,
                 hpx::threads::thread_stacksize::default_, hint),
             num_tasks);
         exec.set_hierarchical_threshold(0);
 
+        // Initialize reduction and ensure proper cleanup
         r.init_iteration(0, 0);
         auto on_exit =
```

Comment:
For the asynchronous case, the cleanup should happen only after all asynchronous tasks have finished running.